src/mame/machine/tsconf.cpp

```diff
@@ -0,0 +1,622 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+
+#include "includes/tsconf.h"
+
+#define W0_RAM (BIT(MEM_CONFIG, 3))
+#define NW0_MAP (BIT(MEM_CONFIG, 2))
+#define W0_WE (BIT(MEM_CONFIG, 1))
+#define ROM128 (BIT(MEM_CONFIG, 0))
+#define RRES (BIT(V_CONFIG, 6, 2))
+
+#define VM static_cast<v_mode>(BIT(V_CONFIG, 0, 2))
+
+static constexpr rectangle resolution_info[5] = {
+	rectangle(52, 256 + 51, 48, 192 + 47), // 52|256|52 x 48-192-48
+	rectangle(20, 320 + 19, 44, 200 + 43), // 20|320|20 x 44-200-44
+	rectangle(20, 320 + 19, 24, 240 + 23), // 20|320|20 x 24-240-24
+	rectangle(0, 360 - 1, 0, 288 - 1),	   //  0|360|0  x  0-288-0
+	rectangle(0, 640 - 1, 0, 240 - 1)	   // text
+};
+
+enum v_mode : u8
+{
+	VM_ZX = 0,
+	VM_16C,
+	VM_256C,
+	VM_TXT
+};
+
+void tsconf_state::tsconf_update_bank1()
+{
+
+	//W0_WE
+	if (NW0_MAP)
+	{
+		m_ROMSelection = PAGE0;
+	}
+	else
+	{
+		/* ROM: 0-SYS, 1-DOS, 2-128, 3-48 */
+		if (m_beta->started() && m_beta->is_active())
+		{
+			m_ROMSelection = ROM128;
+		}
+		else
+		{
+			m_ROMSelection = 0x02 | ROM128;
+		}
+		m_ROMSelection |= (PAGE0 & 0xfc);
+	}
+
+	uint8_t *rom0;
+	if (W0_RAM)
+	{
+		rom0 = m_ram->pointer() + PAGE(m_ROMSelection);
+		m_bank1->set_base(rom0);
+		LOGDEBUG("PAGE0: RAM %X\n", m_ROMSelection);
+	}
+	else
+	{
+		rom0 = &m_p_rom[0x10000 + PAGE(m_ROMSelection & 0x1f)];
+		m_bank1->set_base(rom0);
+		LOGDEBUG("PAGE0: ROM %X\n", m_ROMSelection & 0x1f);
+	}
+	m_ram_0000 = W0_WE ? rom0 : nullptr;
+}
+
+void tsconf_state::tsconf_update_video_mode()
+{
+	rectangle resolution = resolution_info[3];
+	u8 *messram = m_ram->pointer();
+	switch (VM)
+	{
+	case VM_TXT: // Text Mode
+		resolution = resolution_info[4];
+		m_gfxdecode->gfx(1)->set_source(messram + PAGE(V_PAGE ^ 0x01));
+		break;
+	case VM_ZX: // Zx
+	{
+		m_screen_location = messram + ((m_port_7ffd_data & 8) ? PAGE(7) : PAGE(5));
+		break;
+	}
+	default:
+		break;
+	}
+
+	m_screen->configure(resolution.width(), resolution.height(), resolution, HZ_TO_ATTOSECONDS(50));
+
+	// Copy 0xf0 palette to 0x00 for ZX mode
+	u16 palette0_base = VM == VM_ZX ? (0xf0 << 1) : 0x00;
+	for (u8 pen = 0; pen < 16; pen++)
+	{
+		rgb_t rgb = rgb15((m_cram->read(palette0_base + pen * 2 + 1) << 8 | m_cram->read(palette0_base + pen * 2)));
+		subdevice<palette_device>("palette")->set_pen_color(pen, rgb);
+	}
+
+	m_screen->update_now();
```

Comment:
In general you usually want to do this before changing the palette, etc. or you’ll be getting time travel effects where the new palette applies to lines that would have been rendered before the change happened.